File: clients/kotlin/src/main/kotlin/io/lumosdb/VectorClient.kt
package io.lumosdb

import kotlinx.serialization.json.JsonElement
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

/**
 * 向量客户端，处理向量操作
 */
class VectorClient(private val client: LumosDBClient) {
    private val api = ApiClient(client)

    /**
     * 搜索向量
     *
     * @param collectionName 集合名称
     * @param vector 查询向量
     * @param options 搜索选项
     * @return 搜索结果
     */
    suspend fun search(
        collectionName: String,
        vector: List<Float>,
        options: VectorSearchOptions? = null
    ): VectorSearchResponse {
        logger.debug { "Searching vectors in collection $collectionName" }
        
        val request = VectorSearchRequest(
            vector = vector,
            topK = options?.topK,
            scoreThreshold = options?.scoreThreshold,
            filter = options?.filter,
            namespace = options?.namespace
        )
        
        return api.post<VectorSearchRequest, VectorSearchResponse>(
            "/collections/$collectionName/vectors/search",
            request
        )
    }

    /**
     * 添加向量
     *
     * @param collectionName 集合名称
     * @param id 向量ID
     * @param vector 向量数据
     * @param metadata 元数据
     * @param namespace 命名空间
     */
    suspend fun add(
        collectionName: String,
        id: String,
        vector: List<Float>,
        metadata: Map<String, JsonElement>? = null,
        namespace: String? = null
    ) {
        logger.debug { "Adding vector to collection $collectionName with ID $id" }
        
        val request = VectorAddRequest(
            id = id,
            vector = vector,
            metadata = metadata,
            namespace = namespace
        )
        
        api.post<VectorAddRequest, Unit>(
            "/collections/$collectionName/vectors",
            request
        )
    }

    /**
     * 更新向量
     *
     * @param collectionName 集合名称
     * @param id 向量ID
     * @param vector 向量数据
     * @param metadata 元数据
     * @param namespace 命名空间
     */
    suspend fun update(
        collectionName: String,
        id: String,
        vector: List<Float>? = null,
        metadata: Map<String, JsonElement>? = null,
        namespace: String? = null
    ) {
        logger.debug { "Updating vector in collection $collectionName with ID $id" }
        
        val request = VectorUpdateRequest(
            id = id,
            vector = vector,
            metadata = metadata,
            namespace = namespace
        )
        
        api.post<VectorUpdateRequest, Unit>(
            "/collections/$collectionName/vectors/update",
            request
        )
    }

    /**
     * 删除向量
     *
     * @param collectionName 集合名称
     * @param id 向量ID
     * @param namespace 命名空间
     */
    suspend fun delete(
        collectionName: String,
        id: String,
        namespace: String? = null
    ) {
        logger.debug { "Deleting vector from collection $collectionName with ID $id" }
        
        val request = VectorDeleteRequest(
            id = id,
            namespace = namespace
        )
        
        api.post<VectorDeleteRequest, Unit>(
            "/collections/$collectionName/vectors/delete",
            request
        )
    }
}

File: clients/rust/src/vector_client.rs
use serde::de::DeserializeOwned;
use std::sync::Arc;

use crate::api_client::ApiClient;
use crate::error::Result;
use crate::types::{
    ApiResponse, VectorDeleteRequest, VectorInsertRequest, VectorSearchOptions, VectorSearchRequest,
    VectorSearchResponse, VectorUpdateRequest, VectorSearchResult,
};

/// 向量客户端
pub struct VectorClient {
    /// API客户端
    api_client: Arc<ApiClient>,
    /// 集合名称
    collection: String,
}

impl VectorClient {
    /// 创建新的向量客户端
    pub fn new(api_client: Arc<ApiClient>, collection: String) -> Self {
        Self {
            api_client,
            collection,
        }
    }

    /// 构建向量请求路径
    fn build_vector_path(&self, endpoint: &str) -> String {
        format!("/api/vector/collections/{}/{}", self.collection, endpoint)
    }

    /// 向量搜索
    pub async fn search(&self, vector: Vec<f32>, options: Option<VectorSearchOptions>) -> Result<VectorSearchResult> {
        let options = options.unwrap_or_default();
        let request = VectorSearchRequest {
            vector,
            options,
        };
        
        let path = format!("/api/vector/collections/{}/search", self.collection);
        let response: serde_json::Value = self.api_client.post(&path, &request).await?;
        
        if let Some(data) = response.get("data") {
            let result: VectorSearchResult = serde_json::from_value(data.clone())
                .map_err(|e| crate::error::Error::ParseError(e.to_string()))?;
            Ok(result)
        } else {
            Err(crate::error::Error::ParseError("无法解析搜索结果".to_string()))
        }
    }

    /// 向量插入
    pub async fn insert(
        &self,
        id: String,
        vector: Vec<f32>,
        metadata: Option<std::collections::HashMap<String, serde_json::Value>>,
        namespace: Option<String>,
    ) -> Result<()> {
        let request = VectorInsertRequest {
            id,
            vector,
            metadata,
            namespace,
        };
        
        let path = format!("/api/vector/collections/{}/vectors", self.collection);
        let _: serde_json::Value = self.api_client.post(&path, &request).await?;
        Ok(())
    }

    /// 向量更新
    pub async fn update(
        &self,
        id: String,
        vector: Option<Vec<f32>>,
        metadata: Option<std::collections::HashMap<String, serde_json::Value>>,
        namespace: Option<String>,
    ) -> Result<()> {
        let request = VectorUpdateRequest {
            id: id.clone(),
            vector,
            metadata,
            namespace,
        };
        
        let path = format!("/api/vector/collections/{}/vectors/{}", self.collection, id);
        let _: serde_json::Value = self.api_client.put(&path, &request).await?;
        Ok(())
    }

    /// 向量删除
    pub async fn delete(&self, id: String, namespace: Option<String>) -> Result<()> {
        let path = format!("/api/vector/collections/{}/vectors/{}", self.collection, id);
        let _: serde_json::Value = self.api_client.delete(&path).await?;
        Ok(())
    }
} 